File: src/my_first_activity_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;

class NumberCounterNode : public rclcpp::Node
{   
    public:
        NumberCounterNode() : Node("number_counter")
        {
            RCLCPP_INFO(this->get_logger(), "Number counter has been started.");
            subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
                "number", 10,
                std::bind(&NumberCounterNode::callbackNumber, this, _1));
            
            publisher_ = this->create_publisher<example_interfaces::msg::String>("number_count", 10);

            server_ = this->create_service<example_interfaces::srv::SetBool>(
                "reset_counter",
                std::bind(&NumberCounterNode::resetCountCallback, this,
                _1, _2));
        }
        
    private:
        rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
        rclcpp::Publisher<example_interfaces::msg::String>::SharedPtr publisher_;
        int64_t count_ = 0;
        rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;

        void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg)
        {
            
            auto str_msg = example_interfaces::msg::String();
            count_ += msg->data;
            str_msg.data = std::string("data: ")
                + std::to_string(count_);
            publisher_->publish(str_msg);
            RCLCPP_INFO(this->get_logger(), "data: %ld", count_);
        }

        void resetCountCallback(
            const example_interfaces::srv::SetBool::Request::SharedPtr request,
            const example_interfaces::srv::SetBool::Response::SharedPtr response)
        {
            if (request->data) {
                this->count_ = 0;
                response->success = true;
                response->message = "Count has been reset to zero.";
                RCLCPP_INFO(this->get_logger(), "Count has been reset to zero.");
            } else {
                response->success = false;
                response->message = "Count reset not performed.";
                RCLCPP_INFO(this->get_logger(), "Count reset not performed.");
            }
        }
};
int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/my_first_activity_py_pkg/my_first_activity_py_pkg/number_counter.py
#! /usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
from std_msgs.msg import String

class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")
        self.subscriber_ = self.create_subscription(Int64, "number", self.callback_number, 10)
        self.publisher_ = self.create_publisher(String, "number_count", 10)

        self.service_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
        self.get_logger().info("Number counter has been started.")
        self.count_ = 0

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data:
            self.count_ = 0
            response.success = True
            response.message = "Counter has been reset to zero."
            self.get_logger().info("Counter has been reset to zero.")
        else:
            response.success = False
            response.message = "Counter reset not performed."
            self.get_logger().info("Counter reset not performed.")
        return response

    def callback_number(self, msg: Int64):
        str_msg = String()
        self.count_ += msg.data
        str_msg.data = f"data: {self.count_} "
        self.publisher_.publish(str_msg)
        self.get_logger().info(f"data: {self.count_} ")

def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__ == "__main__":
    main()